mkfs/dump.c

```diff
@@ -123,23 +170,54 @@ static u64 get_fs_offset(descriptor fd)
     return fs_offset;
 }
 
+static void usage(const char *prog)
+{
+    const char *p = strrchr(prog, '/');
+    p = p != NULL ? p + 1 : prog;
+    fprintf(stderr, "Usage:\n");
+    fprintf(stderr, "\t%s -d <target dir> <fs image>\n", p);
+    fprintf(stderr, "\t%s -t <fs image>\n", p);
+    fprintf(stderr, "\n");
+    fprintf(stderr, "  -d <target dir>\tCopy filesystem contents from "
+            "<fs image> into <target dir>\n");
+    fprintf(stderr, "  -t\t\t\tDisplay filesystem from <fs image> as a tree\n");
+    exit(EXIT_FAILURE);
+}
+
 int main(int argc, char **argv)
 {
-    if (argc < 3) {
-        const char *p = strrchr(argv[0], '/');
-	p = p != NULL ? p + 1 : argv[0];
-        fprintf(stderr, "usage: %s <fs image> <target dir>\n", p);
-        exit(EXIT_FAILURE);
+    const char *target_dir = NULL;
+    int c;
+
+    while ((c = getopt(argc, argv, "d:t")) != EOF) {
+        switch (c) {
+        case 'd':
+            target_dir = optarg;
+            break;
+        case 't':
+            break;
+        default:
+            usage(argv[0]);
+        }
     }
+    if (optind == argc)
+        usage(argv[0]);
 
-    int fd = open(argv[1], O_RDONLY);
+    int fd = open(argv[optind], O_RDONLY);
     if (fd < 0) {
-        fprintf(stderr, "couldn't open file %s: %s\n", argv[1], strerror(errno));
+        fprintf(stderr, "couldn't open file %s: %s\n", argv[optind],
+            strerror(errno));
         exit(EXIT_FAILURE);
     }
 
     heap h = init_process_runtime();
     tuple root = allocate_tuple();
+    filesystem_complete completion;
+    if (target_dir)
```

Comment:
Do these options need to be mutually exclusive? If -d is specified, a -t will silently have no effect.

Reply:
Indeed the two options don't have to be mutually exclusive. I changed the code so that it always uses the fsc closure, which now takes an additional parameter `options`, and does everything that has been specified from the command line options.